src/providers/data_provider/dp_target_id.c

```diff
@@ -448,10 +448,229 @@ static void dp_req_initgr_pp_set_initgr_timestamp(struct dp_initgr_ctx *ctx,
     }
 }
 
+
+struct dp_sr_resolve_groups_state {
+    struct data_provider *provider;
+    struct dp_initgr_ctx *initgroups_ctx;
+    struct dp_reply_std reply;
+
+    uint32_t *resolve_gids; /* Groups needing resolution */
+    int resolve_gnum;
+    int num_iter;
+    uint32_t gnum;
+};
+
+static errno_t dp_sr_resolve_groups_check(struct dp_sr_resolve_groups_state *state);
+static errno_t dp_sr_resolve_groups_next(struct tevent_req *req);
+static void dp_sr_resolve_groups_done(struct tevent_req *subreq);
+
+struct tevent_req *
+dp_sr_resolve_groups_send(TALLOC_CTX *mem_ctx,
+                 struct tevent_context *ev,
+                 struct dp_reply_std reply,
+                 struct data_provider *provider,
+                 struct dp_initgr_ctx *initgr_ctx)
+{
+
+    struct dp_sr_resolve_groups_state *state;
+    struct tevent_req *req;
+    int ret;
+
+    req = tevent_req_create(mem_ctx, &state, struct dp_sr_resolve_groups_state);
+    if (req == NULL) {
+        DEBUG(SSSDBG_CRIT_FAILURE, "Unable to create tevent request!\n");
+        return NULL;
+    }
+
+    if (initgr_ctx->username == NULL) {
+        ret = EOK;
+        goto done;
+    }
+
+	state->provider = provider;
+	state->initgroups_ctx = initgr_ctx;
+	state->reply = reply;
+	state->gnum = initgr_ctx->gnum;
+
+    /* Check if group is intermediate(has gidNumber and isPosix == False) */
+    state->resolve_gids = talloc_zero_array(state, uint32_t, initgr_ctx->gnum + 1);
+    if (state->resolve_gids == NULL) {
+        ret = ENOMEM;
+        goto done;
+    }
+
+    ret = dp_sr_resolve_groups_check(state);
+    if (ret != EOK) {
+        DEBUG(SSSDBG_OP_FAILURE, "Failed checking groups to resolve\n");
+        goto done;
+    }
+
+	state->num_iter = 0;
+	ret = dp_sr_resolve_groups_next(req);
+	if (ret == EAGAIN) {
+        /* async processing */
+        return req;
+	}
+
+done:
+    if (ret == EOK) {
+        tevent_req_done(req);
+    } else {
+        tevent_req_error(req, ret);
+    }
+    tevent_req_post(req, ev);
+
+    return req;
+}
+
+static errno_t dp_sr_resolve_groups_next(struct tevent_req *req)
+{
+    struct dp_sr_resolve_groups_state *state;
+    struct tevent_req *subreq;
+    struct dp_id_data *ar;
+    uint32_t gid;
+
+    state = tevent_req_data(req, struct dp_sr_resolve_groups_state);
+
+    if (state->num_iter >= state->resolve_gnum) {
+        return EOK;
+    }
+
+    gid = state->resolve_gids[state->num_iter];
+
+    ar = talloc(state, struct dp_id_data);
+    if (!ar) {
```

Comment:
It is generally better to check if (ar == NULL) since pointer is not boolean. It is also required by our coding style:
https://sssd.io/contrib/coding-style.html

```
    /* Do not use !pointer */
    if (pointer == NULL) {
        pointer = "Always use brackets, even for one liner"
    }
```

Reply:
Fixed.